### ISSUE TITLE
Get all tokens including optional (?) tokens from GetExpectedTokens()

### ISSUE THREAD

**Issue by Nickolay-Suslov** [antlr/antlr4#1283]
It seems that GetExpectedTokens() return tokens but not include in list optional tokens. How I can get all this tokens?

**Comment by parrt** [comment]
Hi. got a simple example to explain what is wrong?

**Comment by parrt** [comment]
@Nickolay-Suslov got a simple example?

**Comment by jkra** [comment]
is it solved? I have the same problem, but using antlr-cpp...

here is an example with a simple grammar:

optionalDecl
    : 'hello' ('optional')? ';'
    ;

with the input "hello " I get a syntax error (missing ';') -> ok, but in my error handler (based on BaseErrorListener::syntaxError()) I would expect that parser-getExpectedTokens() gives me 'optional' as expected token too (beside ';') ... but it does not...

or am I wrong?

**Comment by jkra** [comment]
after some googleing it seems that with 

optionalDecl
    : 'hello' ('optional' | ) ';'
    ;

it works -> antlr4 is such a great tool :)

**Comment by parrt** [comment]
See https://github.com/antlr/antlr4/issues/1428 for a new example.

**Comment by parrt** [comment]
Ok, so this is not an error it turns out. It's an issue of *when* you call that function.  before version 4.6, for `('optional')?` we did not detect the missing ';' at the same time we did for `('optional'|)`. 4.6 checks them consistently now. The problem in the first case is that you have asked for the expected tokens when the parser has already moved to the spot where it is only expecting a `;`.  For efficiency, we treat `(A)?` as `A` or *anything*.
